File: parser.py
import re


# Pattern pour supprimer les emojis
_EMOJI_RE = re.compile(
    "["
    "\U0001f600-\U0001f64f"  # emoticons
    "\U0001f300-\U0001f5ff"  # symbols & pictographs
    "\U0001f680-\U0001f6ff"  # transport & map
    "\U0001f1e0-\U0001f1ff"  # flags
    "\U0001f900-\U0001f9ff"  # supplemental symbols
    "\U0001fa00-\U0001fa6f"  # chess symbols
    "\U0001fa70-\U0001faff"  # symbols extended-A
    "\U00002702-\U000027b0"  # dingbats
    "\U0000fe00-\U0000fe0f"  # variation selectors
    "\U0000200d"             # zero width joiner
    "\U000020e3"             # combining enclosing keycap
    "\U00002600-\U000026ff"  # misc symbols
    "\U0000231a-\U0000231b"
    "\U00002934-\U00002935"
    "\U000025aa-\U000025ab"
    "\U000025fb-\U000025fe"
    "\U00002b05-\U00002b07"
    "\U00002b1b-\U00002b1c"
    "\U00002b50"
    "\U00002b55"
    "\U00003030"
    "\U0000303d"
    "\U00003297"
    "\U00003299"
    "]+",
    flags=re.UNICODE,
)


def _clean_emojis(text: str) -> str:
    """Supprime les emojis et espaces superflus."""
    return _EMOJI_RE.sub("", text).strip()


def parse_lead_card(card) -> dict:
    """Extrait les champs d'une carte de lead sur la page de résultats."""
    lead = {}

    # Nom complet (nettoyé des emojis)
    try:
        name_span = card.query_selector("span[data-anonymize='person-name']")
        if name_span:
            full_name = name_span.inner_text().strip()
            parts = full_name.split(" ", 1)
            lead["prenom"] = _clean_emojis(parts[0]) if parts else ""
            lead["nom"] = _clean_emojis(parts[1]) if len(parts) > 1 else ""
    except Exception:
        pass

    # URL profil (lien parent du nom)
    try:
        profile_link = card.query_selector(
            "a[data-control-name='view_lead_panel_via_search_lead_name']"
        )
        if profile_link:
            href = profile_link.get_attribute("href")
            if href:
                lead["url_profil"] = _to_absolute(href)
    except Exception:
        pass

    # Headline (ligne sous le nom = subtitle complet)
    try:
        subtitle_el = card.query_selector(".artdeco-entity-lockup__subtitle")
        if subtitle_el:
            lead["headline"] = subtitle_el.inner_text().strip()
    except Exception:
        pass

    # Description (section "À propos")
    try:
        blurb_el = card.query_selector("[data-anonymize='person-blurb']")
        if blurb_el:
            lead["description"] = blurb_el.inner_text().strip()
    except Exception:
        pass

    # Titre / Poste + Entreprise
    try:
        title_el = card.query_selector("[data-anonymize='title']")
        if title_el:
            raw = title_el.inner_text().strip()
            for sep in [" at ", " chez ", " @ "]:
                if sep in raw:
                    parts = raw.split(sep, 1)
                    lead["titre"] = parts[0].strip()
                    lead["entreprise"] = parts[1].strip()
                    break
            else:
                lead["titre"] = raw
    except Exception:
        pass

    # Entreprise (sélecteur dédié, si pas déjà trouvée)
    if "entreprise" not in lead:
        try:
            company_el = card.query_selector(
                "a[data-anonymize='company-name'], [data-anonymize='company-name']"
            )
            if company_el:
                lead["entreprise"] = company_el.inner_text().strip()
        except Exception:
            pass

    # URL entreprise (format linkedin.com/company/{id})
    try:
        company_link = card.query_selector("a[data-anonymize='company-name']")
        if company_link:
            href = company_link.get_attribute("href") or ""
            match = re.search(r"/company/(\d+)", href)
            if match:
                lead["url_entreprise"] = f"https://www.linkedin.com/company/{match.group(1)}"
    except Exception:
        pass

    # Localisation
    try:
        loc_el = card.query_selector("[data-anonymize='location']")
        if loc_el:
            lead["localisation"] = loc_el.inner_text().strip()
    except Exception:
        pass

    # Tenure et startedOn (depuis le metadata "X ans à ce poste / dans l'entreprise")
    try:
        metadata_el = card.query_selector(".artdeco-entity-lockup__metadata")
        if metadata_el:
            raw = metadata_el.inner_text()
            # Extraire tenure at position (ex: "2 ans 3 mois" ou "3 mois")
            pos_match = re.search(
                r"((?:\d+\s*ans?)?\s*(?:\d+\s*mois)?)\s*à ce poste", raw
            )
            if pos_match and pos_match.group(1).strip():
                lead["tenureAtPosition"] = pos_match.group(1).strip()
            # Extraire tenure at company
            comp_match = re.search(
                r"((?:\d+\s*ans?)?\s*(?:\d+\s*mois)?)\s*dans l.entreprise", raw
            )
            if comp_match and comp_match.group(1).strip():
                lead["tenureAtCompany"] = comp_match.group(1).strip()
            # startedOn : extraire la date si présente (format "depuis mois année")
            start_match = re.search(
                r"depuis\s+([\w]+\s+\d{4})", raw
            )
            if start_match:
                lead["startedOn"] = start_match.group(1)
    except Exception:
        pass

    return lead


def parse_lead_detail(page) -> dict:
    """Extrait les infos détaillées depuis la page profil d'un lead."""
    detail = {}

    # Dernière éducation — année
    try:
        edu_section = page.query_selector(
            "section.education, [data-x--education-history]"
        )
        if edu_section:
            date_els = edu_section.query_selector_all(".date-range span, .pv-entity__dates span")
            years = _extract_years_from_elements(date_els)
            if years:
                detail["annee_derniere_education"] = str(max(years))
    except Exception:
        pass

    # Expérience — première et dernière année
    try:
        exp_section = page.query_selector(
            "section.experience, [data-x--experience]"
        )
        if exp_section:
            date_els = exp_section.query_selector_all(".date-range span, .pv-entity__dates span")
            years = _extract_years_from_elements(date_els)
            if years:
                detail["annee_debut_experience"] = str(min(years))
                detail["annee_fin_experience"] = str(max(years))
    except Exception:
        pass

    # Domaine entreprise
    try:
        industry_el = page.query_selector(
            "[data-anonymize='industry'], .industry, .top-card-layout__entity-info-container .industry"
        )
        if industry_el:
            detail["domaine_entreprise"] = industry_el.inner_text().strip()
    except Exception:
        pass

    return detail


def _extract_years_from_elements(elements) -> list[int]:
    """Extrait les années (4 chiffres) depuis une liste d'éléments DOM."""
    years = []
    for el in elements:
        try:
            text = el.inner_text()
            found = re.findall(r"\b(19|20)\d{2}\b", text)
            years.extend(int(y) for y in found)
        except Exception:
            pass
    return years


def _to_absolute(href: str) -> str:
    """Convertit un href relatif en URL absolue LinkedIn."""
    if href.startswith("http"):
        return href
    return f"https://www.linkedin.com{href}"

File: extension/parser.js
// Port de parser.py — extraction DOM des leads Sales Navigator

const _EMOJI_RE = new RegExp(
  "[" +
    "\u{1F600}-\u{1F64F}" + // emoticons
    "\u{1F300}-\u{1F5FF}" + // symbols & pictographs
    "\u{1F680}-\u{1F6FF}" + // transport & map
    "\u{1F1E0}-\u{1F1FF}" + // flags
    "\u{1F900}-\u{1F9FF}" + // supplemental symbols
    "\u{1FA00}-\u{1FA6F}" + // chess symbols
    "\u{1FA70}-\u{1FAFF}" + // symbols extended-A
    "\u{2702}-\u{27B0}" + // dingbats
    "\u{FE00}-\u{FE0F}" + // variation selectors
    "\u{200D}" + // zero width joiner
    "\u{20E3}" + // combining enclosing keycap
    "\u{2600}-\u{26FF}" + // misc symbols
    "\u{231A}-\u{231B}" +
    "\u{2934}-\u{2935}" +
    "\u{25AA}-\u{25AB}" +
    "\u{25FB}-\u{25FE}" +
    "\u{2B05}-\u{2B07}" +
    "\u{2B1B}-\u{2B1C}" +
    "\u{2B50}" +
    "\u{2B55}" +
    "\u{3030}" +
    "\u{303D}" +
    "\u{3297}" +
    "\u{3299}" +
    "]+",
  "gu"
);

function cleanEmojis(text) {
  return text.replace(_EMOJI_RE, "").trim();
}

function toAbsolute(href) {
  if (href.startsWith("http")) return href;
  return "https://www.linkedin.com" + href;
}

function salesNavToLinkedinUrl(url) {
  // Convertit une URL Sales Navigator en URL profil LinkedIn classique
  // Ex: /sales/lead/ACwAACpkZ6QB...,NAME_SEARCH,kNOb?_ntb=... → https://www.linkedin.com/in/ACwAACpkZ6QB...
  const match = url.match(/\/sales\/lead\/([A-Za-z0-9_-]+)/);
  if (match) {
    return "https://www.linkedin.com/in/" + match[1];
  }
  return url.split("?")[0];
}

function extractYearsFromElements(elements) {
  const years = [];
  for (const el of elements) {
    try {
      const text = el.textContent || "";
      const found = text.match(/\b(19|20)\d{2}\b/g);
      if (found) {
        years.push(...found.map(Number));
      }
    } catch (_) {}
  }
  return years;
}

function parseLeadCard(card) {
  const lead = {};

  // Nom complet
  try {
    const nameSpan = card.querySelector("span[data-anonymize='person-name']");
    if (nameSpan) {
      const fullName = nameSpan.textContent.trim();
      const parts = fullName.split(" ", 2);
      lead.prenom = cleanEmojis(parts[0] || "");
      lead.nom = cleanEmojis(parts.slice(1).join(" ") || "");
    }
  } catch (_) {}

  // URL profil
  try {
    const profileLink = card.querySelector(
      "a[data-control-name='view_lead_panel_via_search_lead_name']"
    );
    if (profileLink) {
      const href = profileLink.getAttribute("href");
      if (href) lead.url_profil = salesNavToLinkedinUrl(toAbsolute(href));
    }
  } catch (_) {}

  // Headline
  try {
    const subtitleEl = card.querySelector(".artdeco-entity-lockup__subtitle");
    if (subtitleEl) {
      lead.headline = subtitleEl.textContent.trim();
    }
  } catch (_) {}

  // Description
  try {
    const blurbEl = card.querySelector("[data-anonymize='person-blurb']");
    if (blurbEl) {
      lead.description = blurbEl.textContent.trim();
    }
  } catch (_) {}

  // Titre / Poste + Entreprise
  try {
    const titleEl = card.querySelector("[data-anonymize='title']");
    if (titleEl) {
      const raw = titleEl.textContent.trim();
      let matched = false;
      for (const sep of [" at ", " chez ", " @ "]) {
        if (raw.includes(sep)) {
          const parts = raw.split(sep, 2);
          lead.titre = parts[0].trim();
          lead.entreprise = parts[1].trim();
          matched = true;
          break;
        }
      }
      if (!matched) {
        lead.titre = raw;
      }
    }
  } catch (_) {}

  // Entreprise (sélecteur dédié, si pas déjà trouvée)
  if (!lead.entreprise) {
    try {
      const companyEl = card.querySelector(
        "a[data-anonymize='company-name'], [data-anonymize='company-name']"
      );
      if (companyEl) {
        lead.entreprise = companyEl.textContent.trim();
      }
    } catch (_) {}
  }

  // URL entreprise
  try {
    const companyLink = card.querySelector("a[data-anonymize='company-name']");
    if (companyLink) {
      const href = companyLink.getAttribute("href") || "";
      const match = href.match(/\/company\/(\d+)/);
      if (match) {
        lead.url_entreprise =
          "https://www.linkedin.com/company/" + match[1];
      }
    }
  } catch (_) {}

  // Localisation
  try {
    const locEl = card.querySelector("[data-anonymize='location']");
    if (locEl) {
      lead.localisation = locEl.textContent.trim();
    }
  } catch (_) {}

  // Tenure et startedOn
  try {
    const metadataEl = card.querySelector(
      ".artdeco-entity-lockup__metadata"
    );
    if (metadataEl) {
      const raw = metadataEl.textContent;
      const posMatch = raw.match(
        /((?:\d+\s*ans?)?\s*(?:\d+\s*mois)?)\s*à ce poste/
      );
      if (posMatch && posMatch[1].trim()) {
        lead.tenureAtPosition = posMatch[1].trim();
      }
      const compMatch = raw.match(
        /((?:\d+\s*ans?)?\s*(?:\d+\s*mois)?)\s*dans l.entreprise/
      );
      if (compMatch && compMatch[1].trim()) {
        lead.tenureAtCompany = compMatch[1].trim();
      }
      const startMatch = raw.match(/depuis\s+([\w]+\s+\d{4})/);
      if (startMatch) {
        lead.startedOn = startMatch[1];
      }
    }
  } catch (_) {}

  return lead;
}

function parseLeadDetail(doc) {
  const detail = {};

  // Dernière éducation — année
  try {
    const eduSection = doc.querySelector(
      "section.education, [data-x--education-history]"
    );
    if (eduSection) {
      const dateEls = eduSection.querySelectorAll(
        ".date-range span, .pv-entity__dates span"
      );
      const years = extractYearsFromElements(dateEls);
      if (years.length) {
        detail.annee_derniere_education = String(Math.max(...years));
      }
    }
  } catch (_) {}

  // Expérience — première et dernière année
  try {
    const expSection = doc.querySelector(
      "section.experience, [data-x--experience]"
    );
    if (expSection) {
      const dateEls = expSection.querySelectorAll(
        ".date-range span, .pv-entity__dates span"
      );
      const years = extractYearsFromElements(dateEls);
      if (years.length) {
        detail.annee_debut_experience = String(Math.min(...years));
        detail.annee_fin_experience = String(Math.max(...years));
      }
    }
  } catch (_) {}

  // Domaine entreprise
  try {
    const industryEl = doc.querySelector(
      "[data-anonymize='industry'], .industry, .top-card-layout__entity-info-container .industry"
    );
    if (industryEl) {
      detail.domaine_entreprise = industryEl.textContent.trim();
    }
  } catch (_) {}

  return detail;
}
